csharp/src/Ice/CollocatedRequestHandler.cs

```diff
@@ -20,143 +22,90 @@ internal class CollocatedRequestHandler : IRequestHandler
-                    if (requestId > 0)
-                    {
-                        _asyncRequests.Remove(requestId);
-                    }
-                    _sendAsyncRequests.Remove(outAsync);
-                    if (outAsync.Exception(ex))
-                    {
-                        Task.Run(outAsync.InvokeException);
-                    }
-                    _adapter.DecDirectCount(); // invokeAll won't be called, decrease the direct count.
-                    return;
-                }
-                if (outAsync is OutgoingAsync o)
-                {
-                    Debug.Assert(o != null);
-                    foreach (KeyValuePair<int, OutgoingAsyncBase> e in _asyncRequests)
-                    {
-                        if (e.Value == o)
-                        {
-                            _asyncRequests.Remove(e.Key);
-                            if (outAsync.Exception(ex))
-                            {
-                                Task.Run(outAsync.InvokeException);
-                            }
-                            return;
-                        }
-                    }
-                }
-            }
-        }
-
         public Connection? GetConnection() => null;
 
-        public void InvokeAsyncRequest(ProxyOutgoingAsyncBase outAsync, bool synchronous)
+        public ValueTask<Task<IncomingResponseFrame>?> SendRequestAsync(OutgoingRequestFrame outgoingRequestFrame,
+            bool oneway, bool synchronous, IInvocationObserver? observer)
         {
             //
             // Increase the direct count to prevent the thread pool from being destroyed before
             // invokeAll is called. This will also throw if the object adapter has been deactivated.
             //
             _adapter.IncDirectCount();
 
+            IChildInvocationObserver? childObserver = null;
             int requestId = 0;
-            try
+            lock (this)
             {
-                lock (this)
+                if (!oneway)
                 {
-                    outAsync.Cancelable(this); // This will throw if the request is canceled
+                    requestId = ++_requestId;
+                }
 
-                    if (!outAsync.IsOneway)
-                    {
-                        requestId = ++_requestId;
-                        _asyncRequests.Add(requestId, outAsync);
-                    }
+                if (observer != null)
+                {
+                    childObserver = observer.GetCollocatedObserver(_adapter, requestId, outgoingRequestFrame.Size);
+                    childObserver?.Attach();
+                }
 
-                    _sendAsyncRequests.Add(outAsync, requestId);
+                if (_adapter.Communicator.TraceLevels.Protocol >= 1)
+                {
+                    ProtocolTrace.TraceCollocatedFrame(_adapter.Communicator, (byte)Ice1Definitions.FrameType.Request,
+                        requestId, outgoingRequestFrame);
```

Comment:
We could get rid of the lock and just use `Interlocked.Increment` for `_requestId`... however this implies that tracing could show traces with out-of-order request IDs... same for the instrumentation observers.

Reply:
I see, I think is nice to see reques id in order, do you think that would be worth caching the protocol trace levels to avoid `_adapter.Communicator.TraceLevels.Protocol` ?

Reply:
I believe we decided to no longer cache this to avoid having too many data members (it used to be a data member).